wgengine/netstack/netstack.go

```diff
@@ -343,74 +343,72 @@ func (ns *Impl) UpdateNetstackIPs(nm *netmap.NetworkMap) {
 			// ours to delete.
 			continue
 		}
-		oldIPs[ap] = true
+		p := netip.PrefixFrom(ip, ap.PrefixLen)
+		oldPfx[p] = true
 	}
-	newIPs := make(map[tcpip.AddressWithPrefix]bool)
+	newPfx := make(map[netip.Prefix]bool)
 
-	isAddr := map[netip.Prefix]bool{}
 	if selfNode.Valid() {
 		for i := range selfNode.Addresses().LenIter() {
-			ipp := selfNode.Addresses().At(i)
-			isAddr[ipp] = true
-			newIPs[ipPrefixToAddressWithPrefix(ipp)] = true
+			p := selfNode.Addresses().At(i)
+			newPfx[p] = true
 		}
-		for i := range selfNode.AllowedIPs().LenIter() {
-			ipp := selfNode.AllowedIPs().At(i)
-			if !isAddr[ipp] && ns.ProcessSubnets {
-				newIPs[ipPrefixToAddressWithPrefix(ipp)] = true
+		if ns.ProcessSubnets {
+			for i := range selfNode.AllowedIPs().LenIter() {
+				p := selfNode.AllowedIPs().At(i)
+				newPfx[p] = true
 			}
 		}
 	}
 
-	ipsToBeAdded := make(map[tcpip.AddressWithPrefix]bool)
-	for ipp := range newIPs {
-		if !oldIPs[ipp] {
-			ipsToBeAdded[ipp] = true
+	pfxToAdd := make(map[netip.Prefix]bool)
+	for p := range newPfx {
+		if !oldPfx[p] {
+			pfxToAdd[p] = true
 		}
 	}
-	ipsToBeRemoved := make(map[tcpip.AddressWithPrefix]bool)
-	for ip := range oldIPs {
-		if !newIPs[ip] {
-			ipsToBeRemoved[ip] = true
+	pfxToRemove := make(map[netip.Prefix]bool)
+	for p := range oldPfx {
+		if !newPfx[p] {
+			pfxToRemove[p] = true
 		}
 	}
 	ns.mu.Lock()
 	for ip := range ns.connsOpenBySubnetIP {
-		ipp := tcpip.AddrFromSlice(ip.AsSlice()).WithPrefix()
-		delete(ipsToBeRemoved, ipp)
+		// TODO(maisem): this looks like a bug, remove or document.
```

Comment:
what looks like a bug here?

Reply:
That we are not cleaning up some addresses. I mostly just want it documented

Reply:
What's "it"? Naman is saying that "this" isn't clear.

Reply:
>That we are not cleaning up some addresses.

It seems like we are leaking them, or removing them accidentally. e.g. if we are advertising `1.2.3.4/32`, and there is an active connection for `1.2.3.4` we won't remove it even if we stop advertising `1.2.3.4`. On the other hand, we might end up removing it when the connection counts drops to 0?